### PR TITLE
Fix: Add MariaDB-specific reflection

### LINKED ISSUE #521: [Bug report] Reflection fails in mariadb
### Describe the bug

In MariaDB, `INFORMATION_SCHEMA.COLUMNS` does not have an `SRS_ID` field.  `geoalchemy2.admin.dialects.mysql.reflect_geometry_column` therefore fails because it assumes this field exists.

### Optional link from https://geoalchemy-2.readthedocs.io which documents the behavior that is expected

_No response_

### To Reproduce

```Python
$ docker run --rm -e MARIADB_ROOT_PASSWORD=root -e MARIADB_PASSWORD=root -p 3306:3306 --name mariadb mariadb:latest
$ docker exec -ti mariadb mariadb -proot
MariaDB> CREATE DATABASE test;
MariaDB> USE test;
MariaDB> CREATE TABLE test_table ( id INTEGER PRIMARY KEY AUTO_INCREMENT, loc POINT );
MariaDB> exit

$ python3 -m venv venv/
$ venv/bin/pip install geoalchemy2 sqlacodegen
$ venv/bin/sqlacodegen mariadb+mariadbconnector://root:root@127.0.0.1:3306/test
```
```


### Error

```Python traceback
/home/tkcook/git/netreach-controller-backend/venv/lib/python3.8/site-packages/sqlacodegen/main.py:47: SAWarning: Did not recognize type 'point' of column 'geolocation'
  metadata.reflect(engine, args.schema, not args.noviews, tables)
Traceback (most recent call last):
  File "/home/tkcook/git/netreach-controller-backend/venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1910, in _execute_context
    self.dialect.do_execute(
  File "/home/tkcook/git/netreach-controller-backend/venv/lib/python3.8/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute
    cursor.execute(statement, parameters)
  File "/home/tkcook/git/netreach-controller-backend/venv/lib/python3.8/site-packages/mariadb/cursors.py", line 310, in execute
    self._readresponse()
mariadb.OperationalError: Unknown column 'SRS_ID' in 'field list'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "venv/bin/sqlacodegen", line 8, in <module>
    sys.exit(main())
  File "/home/tkcook/git/netreach-controller-backend/venv/lib/python3.8/site-packages/sqlacodegen/main.py", line 47, in main
    metadata.reflect(engine, args.schema, not args.noviews, tables)
  File "/home/tkcook/git/netreach-controller-backend/venv/lib/python3.8/site-packages/sqlalchemy/sql/schema.py", line 4901, in reflect
    Table(name, self, **reflect_opts)
  File "<string>", line 2, in __new__
  File "/home/tkcook/git/netreach-controller-backend/venv/lib/python3.8/site-packages/sqlalchemy/util/deprecations.py", line 375, in warned
    return fn(*args, **kwargs)
  File "/home/tkcook/git/netreach-controller-backend/venv/lib/python3.8/site-packages/sqlalchemy/sql/schema.py", line 619, in __new__
    metadata._remove_table(name, schema)
  File "/home/tkcook/git/netreach-controller-backend/venv/lib/python3.8/site-packages/sqlalchemy/util/langhelpers.py", line 70, in __exit__
    compat.raise_(
  File "/home/tkcook/git/netreach-controller-backend/venv/lib/python3.8/site-packages/sqlalchemy/util/compat.py", line 211, in raise_
    raise exception
  File "/home/tkcook/git/netreach-controller-backend/venv/lib/python3.8/site-packages/sqlalchemy/sql/schema.py", line 614, in __new__
    table._init(name, metadata, *args, **kw)
  File "/home/tkcook/git/netreach-controller-backend/venv/lib/python3.8/site-packages/sqlalchemy/sql/schema.py", line 689, in _init
    self._autoload(
  File "/home/tkcook/git/netreach-controller-backend/venv/lib/python3.8/site-packages/sqlalchemy/sql/schema.py", line 724, in _autoload
    conn_insp.reflect_table(
  File "/home/tkcook/git/netreach-controller-backend/venv/lib/python3.8/site-packages/sqlalchemy/engine/reflection.py", line 779, in reflect_table
    self._reflect_column(
  File "/home/tkcook/git/netreach-controller-backend/venv/lib/python3.8/site-packages/sqlalchemy/engine/reflection.py", line 848, in _reflect_column
    table.dispatch.column_reflect(self, table, col_d)
  File "/home/tkcook/git/netreach-controller-backend/venv/lib/python3.8/site-packages/sqlalchemy/event/attr.py", line 247, in __call__
    fn(*args, **kw)
  File "/home/tkcook/git/netreach-controller-backend/venv/lib/python3.8/site-packages/geoalchemy2/admin/__init__.py", line 102, in _reflect_geometry_column
    select_dialect(inspector.bind.dialect.name).reflect_geometry_column(
  File "/home/tkcook/git/netreach-controller-backend/venv/lib/python3.8/site-packages/geoalchemy2/admin/dialects/mysql.py", line 42, in reflect_geometry_column
    geometry_type, srid, nullable_str = inspector.bind.execute(text(geometry_type_query)).one()
  File "/home/tkcook/git/netreach-controller-backend/venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1385, in execute
    return meth(self, multiparams, params, _EMPTY_EXECUTION_OPTS)
  File "/home/tkcook/git/netreach-controller-backend/venv/lib/python3.8/site-packages/sqlalchemy/sql/elements.py", line 334, in _execute_on_connection
    return connection._execute_clauseelement(
  File "/home/tkcook/git/netreach-controller-backend/venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1577, in _execute_clauseelement
    ret = self._execute_context(
  File "/home/tkcook/git/netreach-controller-backend/venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1953, in _execute_context
    self._handle_dbapi_exception(
  File "/home/tkcook/git/netreach-controller-backend/venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 2134, in _handle_dbapi_exception
    util.raise_(
  File "/home/tkcook/git/netreach-controller-backend/venv/lib/python3.8/site-packages/sqlalchemy/util/compat.py", line 211, in raise_
    raise exception
  File "/home/tkcook/git/netreach-controller-backend/venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1910, in _execute_context
    self.dialect.do_execute(
  File "/home/tkcook/git/netreach-controller-backend/venv/lib/python3.8/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute
    cursor.execute(statement, parameters)
  File "/home/tkcook/git/netreach-controller-backend/venv/lib/python3.8/site-packages/mariadb/cursors.py", line 310, in execute
    self._readresponse()
sqlalchemy.exc.OperationalError: (mariadb.OperationalError) Unknown column 'SRS_ID' in 'field list'
[SQL: SELECT DATA_TYPE, SRS_ID, IS_NULLABLE
        FROM INFORMATION_SCHEMA.COLUMNS
        WHERE TABLE_NAME = 'Ap' and COLUMN_NAME = 'geolocation' and table_schema = 'vtba']
(Background on this error at: https://sqlalche.me/e/14/e3q8)
```
```


### Additional context

AFAICT this is a difference between MariaDB and MySQL; MariaDB doesn't support storing SRIDs on spatial types and always does its geometry calculations in Euclidean space.

### GeoAlchemy 2 Version in Use

0.15.2

### Python Version

3.8.19

### Operating system

Linux

### PR DESCRIPTION
**Pull request by tomkcook**
### Description
Fixes: #521 

This resolves #521 -- at least for me -- by adding a MariaDB-specific query for geometry column reflection.

Apologies that I am out of time today to add a test case.  I'll look into it tomorrow.

### Checklist

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [x] A short code fix
	- [x] Please include: `Fixes: #<issue number>` in the description if it solves an existing issue
	  (which must include a complete example of the issue).
	- [ ] Please include tests that fail with the `main` branch and pass with the provided fix.
- [ ] A new feature implementation or update an existing feature
	- [ ] Please include: `Fixes: #<issue number>` in the description if it solves an existing issue
	  (which must include a complete example of the feature).
	- [ ] Please include tests that cover every lines of the new/updated feature.
	- [ ] Please update the documentation to describe the new/updated feature.

### CODE REVIEW

**Comment by tomkcook** [comment]
TBH I struggle a bit to articulate a test that would fail without this change and pass with it, at least within the current test regime.  I can see two options:

* Set up a separate `test_functional_mariadb.py` that replicates most of `test_functional_mysql.py` but with a MariaDB database instead of a MySQL one.
* Write a test that exercises `reflect_geometry_column()` specifically, mock out the engine and check that the executed SQL includes the SRS_ID column if the dialect is MySQL but not if it's MariaDB.

Do the maintainers have a preference for which approach to take?

**Comment by adrien-berchet** [comment]
Hi @tomkcook,
Sorry for the delay, I'm currently on vacation.
Thank you very much for pointing this issue out and for proposing a solution. As far as I can see, your solution looks good but it could be reworked to be a bit more compact and to avoid duplicated code. And for your question, I think that just one test for `reflect_geometry_column` is enough.
I will have some time next week to review this more carefully.

File: geoalchemy2/admin/dialects/mysql.py [hunk 1]
```diff
@@ -31,16 +31,28 @@ def reflect_geometry_column(inspector, table, column_info):
...
+        # Check geometry type, SRID and if the column is nullable
+        geometry_type_query = """SELECT DATA_TYPE, SRS_ID, IS_NULLABLE
+            FROM INFORMATION_SCHEMA.COLUMNS
+            WHERE TABLE_NAME = '{}' and COLUMN_NAME = '{}'""".format(
+            table.name, column_name
+        )
+        if schema is not None:
+            geometry_type_query += """ and table_schema = '{}'""".format(schema)
+        geometry_type, srid, nullable_str = inspector.bind.execute(text(geometry_type_query)).one()
+        is_nullable = str(nullable_str).lower() == "yes"
```
Comment (adrien-berchet) [inline geoalchemy2/admin/dialects/mysql.py:55]:
```suggestion
    if inspector.dialect.name == "mariadb":
        select_srid = "0, "
    else:
        select_srid = "SRS_ID, "

    # Check geometry type, SRID and if the column is nullable
    geometry_type_query = """SELECT DATA_TYPE, {}IS_NULLABLE
        FROM INFORMATION_SCHEMA.COLUMNS
        WHERE TABLE_NAME = '{}' and COLUMN_NAME = '{}'""".format(
        select_srid, table.name, column_name
    )
    if schema is not None:
        geometry_type_query += """ and table_schema = '{}'""".format(schema)
    geometry_type, srid, nullable_str = inspector.bind.execute(text(geometry_type_query)).one()
    is_nullable = str(nullable_str).lower() == "yes"
```
I think it's simpler like this, WDYT?

Comment (tomkcook) [reply geoalchemy2/admin/dialects/mysql.py]: I don't mind either way and it's your code.

Comment (adrien-berchet) [reply geoalchemy2/admin/dialects/mysql.py]: Ahah ok, thanks :)
Do you have some time to work on a test for this soon? Just to know if I have to find some time on my side.

**Comment by adrien-berchet** [comment]
Hi @tomkcook !
I made a few tests in https://github.com/geoalchemy/geoalchemy2/pull/524 and I realized that MariaDB was not as close to MySQL as I expected, so I think it should be processed as a complete different dialect. Nevertheless, there is something I don't understand: I could not find a way to insert a geometry using a WKB string, is it possible? Otherwise we will have to convert all WKB strings into WKT strings, which is quite bad for perf.

**Comment by adrien-berchet** [comment]
Hi @tomkcook !
I think #524 solved most of the issues with MariaDB, could you test it in your workflow to ensure everything is all right please?

**Comment by adrien-berchet** [comment]
This PR was included in #524 so it can be closed now.
Thanks again for your work @tomkcook !

**Comment by tomkcook** [comment]
No worries.  Sorry for not getting back to you sooner.  Thanks for merging the fix.

**Comment by adrien-berchet** [comment]
With pleasure.
Btw, did you have the opportunity to test the last release on MariaDB? I did multiple changes so new issues may raise.

**Pull request closed without merging**